File: src/riptide_turtle/src/turtle_kill_switch.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/bool.hpp"

class TurtleEnablePublisher : public rclcpp::Node
{
public:
    TurtleEnablePublisher()
    : Node("turtle_kill_switch_publisher"), state_(true)
    {
        publisher_ = this->create_publisher<std_msgs::msg::Bool>("turtle_enabled", 10);
        state_timer_ = this->create_wall_timer(
            std::chrono::seconds(30),
            std::bind(&TurtleEnablePublisher::toggle_state_callback, this));
        publish_timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&TurtleEnablePublisher::publish_state_callback, this));

        RCLCPP_INFO(this->get_logger(), "Turtle Enable Publisher started!");
    }

private:
    void toggle_state_callback()
    {
        state_ = !state_;
        auto state_str = state_ ? "Enabled" : "Disabled";
        RCLCPP_INFO(this->get_logger(), "State toggled: %s", state_str);
    }

    void publish_state_callback()
    {
        auto msg = std_msgs::msg::Bool();
        msg.data = state_;
        publisher_->publish(msg);
        auto state_str = state_ ? "Enabled" : "Disabled";
        RCLCPP_INFO(this->get_logger(), "Published: %s", state_str);
    }

    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr state_timer_;
    rclcpp::TimerBase::SharedPtr publish_timer_;
    bool state_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurtleEnablePublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/riptide_turtle/src/turtle_kill_switch.py
#!/usr/bin/env python3

import rclpy  # Import the ROS 2 Python client library
from rclpy.node import Node  # Import the Node class from rclpy
from std_msgs.msg import Bool  # Import the Bool message type for enabling/disabling the turtle

class TurtleEnablePublisher(Node):  # Define a new class 'TurtleEnablePublisher' that inherits from the ROS 2 'Node' class
    def __init__(self):  # The constructor method to initialize the node's attributes and setup
        super().__init__('turtle_kill_switch_publisher')  # Initialize the node with the name 'turtle_kill_switch_publisher'
        
        # TODO: Create a publisher that publishes Bool messages to the 'turtle_enabled' topic
        # self.publisher_ = ...

        # TODO: Set up a timer that periodically calls the toggle_state_callback method every 30 seconds
        # self.state_timer = ...

        # TODO: Set up a timer that periodically calls the publish_state_callback method every second
        # self.publish_timer = ...

        self.state = True  # Initialize the state as True (enabled)
        self.get_logger().info("Turtle Enable Publisher started!")  # Log a message to confirm the node has started

    def toggle_state_callback(self):
        # Toggle the state between True and False
        self.state = not self.state
        state_str = "Enabled" if self.state else "Disabled"
        self.get_logger().info(f"State toggled: {state_str}")

    def publish_state_callback(self):
        # Create and configure a Bool message with the current state
        msg = Bool()
        msg.data = self.state

        # TODO: Publish the Bool message
        # self.publisher_.publish(msg)

        state_str = "Enabled" if self.state else "Disabled"
        self.get_logger().info(f"Published: {state_str}")

def main(args=None):
    rclpy.init(args=args)  # Initialize the ROS 2 Python client library
    node = TurtleEnablePublisher()  # Create an instance of the TurtleEnablePublisher
    rclpy.spin(node)  # Keep the node running until manually interrupted
    node.destroy_node()  # Clean up the node after shutdown
    rclpy.shutdown()  # Shutdown the ROS 2 client library

if __name__ == '__main__':
    main()  # Run the main function if this script is executed directly